File: backend/src/models/RepairTicket.ts
import mongoose, { Schema, Document } from "mongoose";

export type RepairTicketStatus =
  | "new"
  | "assigned"
  | "quoted"
  | "pending_confirm"
  | "waiting_parts"
  | "parts_ready"
  | "customer_rejected"
  | "returned"
  | "repaired"
  | "delivered"
  | "cancelled"
  | "outsourced";

export type ServiceType = "warranty" | "warranty_repair" | "service";
export type ServiceLocation = "at_station" | "at_home";

export interface RepairTicketDocument extends Document {
  ticketNumber: string;
  ticketRefNumber?: string;
  status: RepairTicketStatus;
  serviceType: ServiceType;
  serviceLocation: ServiceLocation;
  isUrgent: boolean;

  // Thông tin sản phẩm
  productName?: string;
  manufacturer?: string;
  modelName?: string;
  serialNumber?: string;
  purchaseDate?: Date;
  purchasePlace?: string;
  faultDescription: string;
  accessories?: string;

  // Thông tin khách hàng
  customerName: string;
  customerPhone: string;
  customerAddress?: string;

  // Thông tin phiếu
  area?: string;
  quotedPrice?: number;
  note?: string;
  internalNote?: string;

  // Tiếp nhận
  receivedDate: Date;
  receivedBy?: string;
  appointmentDate?: Date;
  completedDate?: Date;
  technician?: mongoose.Types.ObjectId;
  outsourcedTo?: string;    // Tên thợ ngoài (nếu giao ra ngoài)

  // Ảnh theo giai đoạn
  intakeImages: string[];      // Ảnh tiếp nhận
  faultImages: string[];       // Ảnh hư hỏng / báo giá
  completedImages: string[];   // Ảnh sửa xong
  statusImages?: Map<string, string[]>; // Ảnh upload bởi KTV theo từng trạng thái

  createdAt: Date;
  updatedAt: Date;
}

const repairTicketSchema = new Schema<RepairTicketDocument>(
  {
    ticketNumber: { type: String, required: true, unique: true, index: true },
    ticketRefNumber: { type: String },
    status: {
      type: String,
      enum: [
        "new", "assigned", "quoted", "pending_confirm",
        "waiting_parts", "parts_ready", "customer_rejected",
        "returned", "repaired", "delivered", "cancelled", "outsourced"
      ],
      default: "new",
      index: true
    },
    serviceType: {
      type: String,
      enum: ["warranty", "warranty_repair", "service"],
      default: "warranty"
    },
    serviceLocation: {
      type: String,
      enum: ["at_station", "at_home"],
      default: "at_station"
    },
    isUrgent: { type: Boolean, default: false },

    productName: { type: String },
    manufacturer: { type: String },
    modelName: { type: String },
    serialNumber: { type: String, index: true },
    purchaseDate: { type: Date },
    purchasePlace: { type: String },
    faultDescription: { type: String, required: true },
    accessories: { type: String },

    customerName: { type: String, required: true },
    customerPhone: { type: String, required: true, index: true },
    customerAddress: { type: String },

    area: { type: String },
    quotedPrice: { type: Number },
    note: { type: String },
    internalNote: { type: String },

    receivedDate: { type: Date, default: Date.now },
    receivedBy: { type: String },
    appointmentDate: { type: Date },
    completedDate: { type: Date },
    technician: { type: Schema.Types.ObjectId, ref: "User" },
    outsourcedTo: { type: String },

    intakeImages: [{ type: String }],
    faultImages: [{ type: String }],
    completedImages: [{ type: String }],
    statusImages: { type: Map, of: [{ type: String }], default: {} }
  },
  { timestamps: true }
);

repairTicketSchema.index({ createdAt: -1 });
repairTicketSchema.index({ customerPhone: 1, createdAt: -1 });

export const RepairTicketModel = mongoose.model<RepairTicketDocument>("RepairTicket", repairTicketSchema);

File: flutter/lib/models/repair_ticket.dart
class RepairTicket {
  final String id;
  final String ticketNumber;
  final String? ticketRefNumber;
  final String status;
  final String serviceType;
  final String serviceLocation;
  final bool isUrgent;

  // Sản phẩm
  final String? productName;
  final String? manufacturer;
  final String? modelName;
  final String? serialNumber;
  final String faultDescription;

  // Khách hàng
  final String customerName;
  final String customerPhone;
  final String? customerAddress;
  final String? area;

  // Thông tin phiếu
  final double? quotedPrice;
  final String? note;
  final String? internalNote;

  final DateTime receivedDate;
  final DateTime? appointmentDate;
  final DateTime? completedDate;

  final List<String> intakeImages;
  final List<String> faultImages;
  final List<String> completedImages;

  RepairTicket({
    required this.id,
    required this.ticketNumber,
    this.ticketRefNumber,
    required this.status,
    required this.serviceType,
    required this.serviceLocation,
    required this.isUrgent,
    this.productName,
    this.manufacturer,
    this.modelName,
    this.serialNumber,
    required this.faultDescription,
    required this.customerName,
    required this.customerPhone,
    this.customerAddress,
    this.area,
    this.quotedPrice,
    this.note,
    this.internalNote,
    required this.receivedDate,
    this.appointmentDate,
    this.completedDate,
    this.intakeImages = const [],
    this.faultImages = const [],
    this.completedImages = const [],
  });

  factory RepairTicket.fromJson(Map<String, dynamic> json) {
    return RepairTicket(
      id: json['id']?.toString() ?? json['_id']?.toString() ?? '',
      ticketNumber: json['ticketNumber']?.toString() ?? '',
      ticketRefNumber: json['ticketRefNumber']?.toString(),
      status: json['status']?.toString() ?? 'new',
      serviceType: json['serviceType']?.toString() ?? 'service',
      serviceLocation: json['serviceLocation']?.toString() ?? 'at_station',
      isUrgent: json['isUrgent'] == true,
      productName: json['productName']?.toString(),
      manufacturer: json['manufacturer']?.toString(),
      modelName: json['modelName']?.toString(),
      serialNumber: json['serialNumber']?.toString(),
      faultDescription: json['faultDescription']?.toString() ?? '',
      customerName: json['customerName']?.toString() ?? '',
      customerPhone: json['customerPhone']?.toString() ?? '',
      customerAddress: json['customerAddress']?.toString(),
      area: json['area']?.toString(),
      quotedPrice: (json['quotedPrice'] as num?)?.toDouble(),
      note: json['note']?.toString(),
      internalNote: json['internalNote']?.toString(),
      receivedDate:
          DateTime.tryParse(json['receivedDate']?.toString() ?? '') ??
              DateTime.now(),
      appointmentDate: json['appointmentDate'] != null
          ? DateTime.tryParse(json['appointmentDate'].toString())
          : null,
      completedDate: json['completedDate'] != null
          ? DateTime.tryParse(json['completedDate'].toString())
          : null,
      intakeImages: List<String>.from(json['intakeImages'] ?? []),
      faultImages: List<String>.from(json['faultImages'] ?? []),
      completedImages: List<String>.from(json['completedImages'] ?? []),
    );
  }
}
